File: frontend/types/live.ts
export type Currency = "INR" | "USD";
export type Category = "accommodation" | "transportation" | "food" | "activities" | "miscellaneous";
export const categories: Category[] = ["accommodation", "transportation", "food", "activities", "miscellaneous"];
export interface Preferences {
  destination: string; number_of_days: number; start_date: string; budget: number | null;
  currency: Currency; interests: string[]; travelers: number; rooms: number;
  transport_mode: "walking" | "driving"; dietary_preferences: string[];
  pace: "relaxed" | "balanced" | "busy"; allowances: Record<Category, number | null>;
}
export interface Source {
  provider: string; status: "live" | "cached" | "unavailable" | "partial" | "estimated";
  retrieved_at: string | null; expires_at: string | null; message: string;
}
export interface Place {
  id: string; name: string; kind: string; latitude: number; longitude: number;
  tags: Record<string, string>; opening_hours: string | null; phone: string | null;
  website: string | null; source: Source; score: number; reason: string;
}
export interface Activity {
  id: string; place: Place; date: string; start_time: string; end_time: string;
  duration_minutes: number; duration_basis: string; locked: boolean; completed: boolean; committed: boolean;
}
export interface Route {
  from_id: string; to_id: string; minutes: number | null; distance_km: number | null;
  traffic_delay_minutes: number | null; instructions: string[]; source: Source;
}
export interface HotelOffer {
  id: string; hotel_id: string; name: string; amount: number; currency: string;
  check_in: string; check_out: string; rooms: number; source: Source;
}
export interface Budget {
  lines: {category: Category; planned: number | null; spent: number; projected: number | null; basis: string}[];
  known_total: number; total: number | null; spent: number; complete: boolean;
  within_budget: boolean | null; remaining: number | null; conversion_note: string | null;
}
export interface LivePlan {
  places: Place[]; restaurants: Place[]; lodging: Place[]; emergency: Place[]; hotels: HotelOffer[];
  weather: {date: string; temperature_min: number | null; temperature_max: number | null; rain_probability: number | null; wind_kmh: number | null; avoid_outdoor: boolean}[];
  routes: Route[]; activities: Activity[]; budget: Budget; sources: Record<string, Source>;
  agents: Record<string, string>; warnings: string[]; conflicts: string[]; valid: boolean;
  attempts: number; timezone: string; generated_at: string;
}
export interface Trip {
  id: string; version: number; preferences: Preferences; plan: LivePlan;
  expenses: {id: string; category: Category; amount: number; description: string; created_at: string}[];
  conversation: {role: "user" | "assistant"; content: string; created_at: string}[];
  monitoring: boolean; last_checked: string | null; next_check: string | null; monitoring_error: string | null;
  created_at: string; updated_at: string;
}
export interface TripSummary {
  id: string; version: number; destination: string; start_date: string;
  number_of_days: number; monitoring: boolean; valid: boolean;
}
export interface Revision {version: number; created_at: string; reason: string; before: Trip | null; after: Trip}
export interface Capabilities {local_model: {configured: boolean; message: string}; hotels_configured: boolean; traffic_configured: boolean}

File: models/live.py
"""Validated contracts for persisted, live-data trips. Money is in trip currency."""
from datetime import date, datetime, timezone
from typing import Literal
from uuid import uuid4

from pydantic import BaseModel, ConfigDict, Field, field_validator


def utcnow() -> str:
    return datetime.now(timezone.utc).isoformat()


class Model(BaseModel):
    model_config = ConfigDict(extra="forbid", allow_inf_nan=False)


Category = Literal["accommodation", "transportation", "food", "activities", "miscellaneous"]


class Allowances(Model):
    """Total trip allowances, including all travelers, in the trip currency."""
    accommodation: float | None = Field(None, ge=0)
    transportation: float | None = Field(None, ge=0)
    food: float | None = Field(None, ge=0)
    activities: float | None = Field(None, ge=0)
    miscellaneous: float | None = Field(None, ge=0)


class Preferences(Model):
    destination: str = Field(min_length=1, max_length=200)
    number_of_days: int = Field(3, ge=1, le=60)
    start_date: date = Field(default_factory=date.today)
    budget: float | None = Field(None, ge=0)
    currency: Literal["INR", "USD"] = "INR"
    interests: list[str] = Field(default_factory=list, max_length=20)
    travelers: int = Field(1, ge=1, le=20)
    rooms: int = Field(1, ge=1, le=10)
    transport_mode: Literal["walking", "driving"] = "walking"
    dietary_preferences: list[str] = Field(default_factory=list, max_length=10)
    pace: Literal["relaxed", "balanced", "busy"] = "balanced"
    allowances: Allowances = Field(default_factory=Allowances)

    @field_validator("destination")
    @classmethod
    def trim_destination(cls, value):
        if not value.strip():
            raise ValueError("Destination is required")
        return value.strip()


class Source(Model):
    provider: str
    status: Literal["live", "cached", "unavailable", "partial", "estimated"] = "unavailable"
    retrieved_at: str | None = None
    expires_at: str | None = None
    message: str = ""


class Place(Model):
    id: str
    name: str
    kind: str
    latitude: float = Field(ge=-90, le=90)
    longitude: float = Field(ge=-180, le=180)
    tags: dict[str, str] = Field(default_factory=dict)
    opening_hours: str | None = None
    phone: str | None = None
    website: str | None = None
    source: Source
    score: float = 0
    reason: str = ""


class HotelOffer(Model):
    id: str
    hotel_id: str
    name: str
    amount: float = Field(ge=0)
    currency: str
    check_in: date
    check_out: date
    rooms: int = Field(ge=1)
    source: Source


class WeatherDay(Model):
    date: date
    temperature_min: float | None = None
    temperature_max: float | None = None
    rain_probability: float | None = None
    wind_kmh: float | None = None
    avoid_outdoor: bool = False


class Route(Model):
    from_id: str
    to_id: str
    minutes: float | None = Field(None, ge=0)
    distance_km: float | None = Field(None, ge=0)
    traffic_delay_minutes: float | None = Field(None, ge=0)
    instructions: list[str] = Field(default_factory=list)
    source: Source


class Activity(Model):
    id: str
    place: Place
    date: date
    start_time: str = Field(pattern=r"^(?:[01]\d|2[0-3]):[0-5]\d$")
    end_time: str = Field(pattern=r"^(?:[01]\d|2[0-3]):[0-5]\d$")
    duration_minutes: int = Field(ge=1)
    duration_basis: str = "Planned visit duration based on your activity pace"
    locked: bool = False
    completed: bool = False
    committed: bool = False


class Expense(Model):
    id: str = Field(default_factory=lambda: str(uuid4()))
    category: Category
    amount: float = Field(ge=0)
    description: str = Field(min_length=1, max_length=300)
    created_at: str = Field(default_factory=utcnow)


class BudgetLine(Model):
    category: Category
    planned: float | None = None
    spent: float = 0
    projected: float | None = None
    basis: str


class Budget(Model):
    lines: list[BudgetLine] = Field(default_factory=list)
    known_total: float = 0
    total: float | None = None
    spent: float = 0
    complete: bool = False
    within_budget: bool | None = None
    remaining: float | None = None
    conversion_note: str | None = None


class LivePlan(Model):
    places: list[Place] = Field(default_factory=list)
    restaurants: list[Place] = Field(default_factory=list)
    lodging: list[Place] = Field(default_factory=list)
    emergency: list[Place] = Field(default_factory=list)
    hotels: list[HotelOffer] = Field(default_factory=list)
    weather: list[WeatherDay] = Field(default_factory=list)
    routes: list[Route] = Field(default_factory=list)
    activities: list[Activity] = Field(default_factory=list)
    budget: Budget = Field(default_factory=Budget)
    sources: dict[str, Source] = Field(default_factory=dict)
    agents: dict[str, str] = Field(default_factory=dict)
    warnings: list[str] = Field(default_factory=list)
    conflicts: list[str] = Field(default_factory=list)
    valid: bool = False
    attempts: int = 0
    timezone: str = "UTC"
    generated_at: str = Field(default_factory=utcnow)


class Message(Model):
    role: Literal["user", "assistant"]
    content: str
    created_at: str = Field(default_factory=utcnow)


class Trip(Model):
    id: str = Field(default_factory=lambda: str(uuid4()))
    version: int = 1
    preferences: Preferences
    plan: LivePlan = Field(default_factory=LivePlan)
    expenses: list[Expense] = Field(default_factory=list)
    conversation: list[Message] = Field(default_factory=list)
    monitoring: bool = True
    last_checked: str | None = None
    next_check: str | None = None
    monitoring_error: str | None = None
    created_at: str = Field(default_factory=utcnow)
    updated_at: str = Field(default_factory=utcnow)
